api/layer/multipart_upload.go

```diff
@@ -0,0 +1,494 @@
+package layer
+
+import (
+	"context"
+	"io"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/nspcc-dev/neofs-api-go/pkg/object"
+	"github.com/nspcc-dev/neofs-api-go/pkg/owner"
+	"github.com/nspcc-dev/neofs-s3-gw/api/data"
+	"github.com/nspcc-dev/neofs-s3-gw/api/errors"
+	"go.uber.org/zap"
+)
+
+const (
+	UploadIDAttributeName         = "S3-Upload-Id"
+	UploadPartNumberAttributeName = "S3-Upload-Part-Number"
+	UploadKeyAttributeName        = "S3-Upload-Key"
+	UploadHasTags                 = "S3-Upload-Has-Tags"
+	UploadHasACL                  = "S3-Upload-Has-ACL"
+	UploadPartKeyPrefix           = ".upload-"
+
+	uploadMinSize = 5 * 1048576       // 5MB
+	uploadMaxSize = 5 * 1099511627776 // 5TB
+)
+
+type (
+	UploadInfoParams struct {
+		UploadID string
+		Bkt      *data.BucketInfo
+		Key      string
+	}
+
+	UploadPartParams struct {
+		Info       *UploadInfoParams
+		PartNumber int
+		Size       int64
+		Reader     io.Reader
+		Header     map[string]string
+	}
+
+	CompleteMultipartParams struct {
+		Info  *UploadInfoParams
+		Parts []*CompletedPart
+	}
+
+	CompletedPart struct {
+		ETag       string
+		PartNumber int
+	}
+
+	Part struct {
+		ETag         string
+		LastModified string
+		PartNumber   int
+		Size         int64
+	}
+
+	ListMultipartUploadsParams struct {
+		Bkt            *data.BucketInfo
+		Delimiter      string
+		EncodingType   string
+		KeyMarker      string
+		MaxUploads     int
+		Prefix         string
+		UploadIdMarker string
+	}
+
+	ListPartsParams struct {
+		Info             *UploadInfoParams
+		MaxParts         int
+		PartNumberMarker int
+	}
+
+	ListPartsInfo struct {
+		Parts                []*Part
+		Owner                *owner.ID
+		NextPartNumberMarker int
+		IsTruncated          bool
+	}
+
+	ListMultipartUploadsInfo struct {
+		Prefixes           []string
+		Uploads            []*UploadInfo
+		IsTruncated        bool
+		NextKeyMarker      string
+		NextUploadIdMarker string
+	}
+	UploadInfo struct {
+		IsDir    bool
+		Key      string
+		UploadID string
+		Owner    *owner.ID
+		Created  time.Time
+	}
+)
+
+func (n *layer) UploadPart(ctx context.Context, p *UploadPartParams) (*data.ObjectInfo, error) {
+	if p.PartNumber != 0 {
+		_, err := n.GetUploadInitInfo(ctx, p.Info)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if p.Size > uploadMaxSize {
```

Comment:
Shouldn't we also check `uploadMinSize`?

Reply:
We don't have limit `uploadMinSize` for the last part. When we just upload a part we don't know if the part the last or not.